### ISSUE TITLE
Improve static HTML rendering API (example: for CSS-in-JS libs)

### ISSUE THREAD

**Issue by hilja** [phenomic/phenomic#1094]
Trying to use [Fela](https://github.com/rofrischmann/fela). It does works in the dev mode `phenomic start` but gives an error when `phenomic build`:

```
An error occured when Phenomic tried to render
An error occured when Phenomic tried to render archive
An error occured when Phenomic tried to render after/:after

/Users/foo/bar/node_modules/.cache/phenomic/webpack/webpack:/~/fela/es/bindings/createComponentFactory.js:28
        throw new Error("createComponent() can't render styles without the renderer in the context. Missing react-fela's <Provider /> at the app root?");
```

[react-fela](https://github.com/rofrischmann/fela/tree/master/packages/react-fela) works so that you wrap the whole app in the fela [Provider](https://github.com/rofrischmann/fela/blob/master/packages/react-fela/docs/Provider.md): `<Provider renderer={renderer}>`. The error is the same as if wouldn't use the `Provider` at all.

Here's the  code I'm testing with:

```
import { createRenderer } from 'fela'
import { Provider } from 'react-fela'

const renderer = createRenderer()

export default () => (
  <Provider renderer={renderer}>
    <Router history={browserHistory}>
      <Route path='/' component={homeContainer} />
      <Route path='/blog/*' component={blogPostContainer} collections='posts' />
      <Route path='/archive' component={archiveContainer} />
      <Route path='/after/:after' component={homeContainer} />
      <Route path='*' component={PageError} />
    </Router>
  </Provider>
)
```

And my component:

```
import { StyleSheet } from 'fela-tools'
import { createComponent } from 'react-fela'
const header = { backgroundColor: 'red' }
const styles = StyleSheet.create({ header })
const HeaderEl = createComponent(styles.header, 'header')

const Header = () => (
  <HeaderEl><h1>foo</h1></HeaderEl>
)
```

I'm also new to Fela. But since it works in dev but not in build, I have a hunch that it might be something with how the Phenomic builds the site.

I'm on the `1.0.0-alpha.3` version.

**Comment by MoOx** [comment]
Did you try using a custom Html component?

**Comment by dlindahl** [comment]
I'm getting errors trying to use [Styletron](http://styletron.js.org/) with `1.0.0-alpha.3` and a custom Html component.

I am also able to get it working in development with `phenomic start` but not `phenomic build`.

I think there are two things going on here that are contributing to this issue:

1. When server rendering, a route is matched and passed to React Router's `RouterContext` which then iterates over the possible routes, and [pulls the appropriate route out of the array of route components based on the matched route's index](https://github.com/ReactTraining/react-router/blob/c45729d0ebc079cf646e8b26774b035a6179e86f/modules/RouterContext.js#L57). The "problem" is that this circumvents the rendering of non-React Router components in the component tree. It will never "match" and non-React Router component.

    Luckily, this can be solved by using an "empty" route and rendering a generic `ContextProvider` that can encapsulate your needs:

    ```jsx
    export default class ContextProvider extends Component {
       // ...
      render() {
        const renderer = createRenderer()
        return (
          <Provider renderer={renderer}>
            {Children.only(this.props.children)}
          </Provider>
        )
      }
    }
    ```

    ```jsx
    // in your routes file
    const SiteRouter = () => (
      <Router history={browserHistory}>
        <Route component={ContextProvider}>
          <Route component={HomeContainer} path="/"/>
        </Route>
      </Router>
    )
    ```

    This allows your provider to be rendered as a route that will always match. This also appears to work with Styletron.

1. The second problem, at least as far as Styletron is concerned, is that the component tree is rendered to a string _before being passed into the Html container_.

    Styletron builds a cache of used styles while building the component tree which are then flushed and injected into the rendered markup by the developer:

    ```jsx
    import Styletron from 'styletron-server';
    import {StyletronProvider} from 'styletron-react';

    function render() {
      const styletron = new Styletron();
      const appMarkup = ReactDOM.renderToString(
        <StyletronProvider styletron={styletron}>
          <App/>
        </StyletronProvider>
      );

      const stylesForHead = styletron.getStylesheetsHtml();
  
      return `<html><head>${stylesForHead}</head><body>${appMarkup}</body></html>`;
    }
    ```
    This doesn't work in Phenomic because the [component tree is rendered to a string before being passed to the Html container](https://github.com/phenomic/phenomic/blob/c7bddecaa5cad5783569343f3b0692dddadc80cb/packages/plugin-renderer-react/src/render/server.js#L38-L50). This means that the instance of Styletron that contains all the CSS rules is lost and `getStylesheetsHtml` can no longer be called to build the stylesheet. This also means that the `StyletronProvider` cannot be moved to the Html container because the component tree has already been rendered and converted to a string by then.

    Aside from the [developer experience and hot reloading support](https://github.com/phenomic/phenomic/blob/c7bddecaa5cad5783569343f3b0692dddadc80cb/packages/plugin-renderer-react/src/server/renderHTML.js#L13-L18), is there a reason why the Html container is not part of the render path?

**Comment by dlindahl** [comment]
I have figured out a way to get Styletron to work with Phenomic, but it requires a small change to Phenomic's code which I think is in the project's best interest anyway:

*tl;dr*
* Phenomic should not stringify it's state (and probably not the body, either) in order to allow for more flexible custom HTML Containers
* Providing a specific state object for developer use would be advantageous, especially when dealing with CSS-in-JS libraries
* The same Html Container should be used during both client _and_ server rendering paths

<details>
<summary>My proposed solution is contained within</summary>
The issue at hand has two parts:

1. As previously stated, the application's body is stringified before being passed to the HTML Container. This prevents any third-party "providers" from wrapping the application.

2. If a third-party provider wraps the application in the routes file (as described in my previous comment), the third-party library's instance disappears once the body is stringified, preventing those libraries from injecting necessary content.

I was able to overcome these limitations by adding the Styletron instance, defined above in my `ContextProvider` to Phenomic's internal data store. Then, instead of stringifying the state before passing it to the Html Container, I modified `renderHTML` to pass the data store object directly into the container component. This allowed me to regain the reference to the Styletron instance and inject the compiled CSS rules:

```jsx
export default class ContextProvider extends Component {
  // ...
  static contextTypes = {
    // ...
    phenomic: object
  }
  __styletron__ = null
  constructor (props, context) {
    super(props, context)
    this.context.phenomic.set('styletron', new Styletron())
  }
  // ...
}
```

```jsx
// in /plugin-renderer-react/src/server/renderHTML.js
const renderHTML: PhenomicPluginRenderHTMLType = (
  config,
  props = {},
  Html = DefaultHtml
) => {
  const html = (
    <Html
      body={
        <div
          id="PhenomicRoot"
          dangerouslySetInnerHTML={{ __html: props.body || "" }}
        />
      }
      state={props.state}
      script={<script src={`/${config.bundleName}.js`} async />}
    />
  );
  return `<!DOCTYPE html>${ReactDOMServer.renderToStaticMarkup(html)}`;
};
```

```jsx
// in my custom Html Container
const HtmlContainer = ({ body, script, state }) => {
  const { styletron, ...phenomicState } = state
  const stylesForHead = styletron.node.getCss()
  return (
    <html>
      <head>
        <style>
          {stylesForHead}
        </style>
      </head>
      <body>
        {body}
        <script
          dangerouslySetInnerHTML={{
            __html: JSON.stringify(phenomicState)
          }}
          id="PhenomicHydration"
          type="text/json"
        />
        {/* Phenomic entry script */}
        {script}
      </body>
    </html>
  )
}
HtmlContainer.propTypes = {
  body: string,
  script: string,
  state: object
}
```

While this works, it is obviously abusing the purpose of Phenomic's data store, but I will get to the in a moment.

Since I am already providing my own custom Html Container, I should be responsible for serializing the body and state and wrapping them in the appropriately ID'd `div`. All it would take is a small change to `renderHTML`, the default Html Container, and the documentation. 

Regarding my abuse of Phenomic's data store, this is a less than ideal solution because at some point, Phenomic runs queries against each key in the store, mutating the object in the process. This means that my `styletron` key gets wrapped in a `node` object and does God knows what to the query runner internals.

It would be nice if Phenomic could expose some sort of blank state object, that remains untouched by Phenomic, that could be used by developers to integrate with other third-party libraries. For lack of a better name, the existing store could be renamed from `phenomic` to `phenomicCollections` and this new store could be named `phenomicStore`. These are not great names, but its at least a starting place.

Lastly, I found it confusing that my custom Html Container was not used at all when rendering on the client. It just so happens that `react-helmet` and Styletron (and presumably Fela as stated in the OP) both work out of the box without any changes, but if I was adding some sort of custom attribute to the `html` tag, a custom footer `script` tag, etc., it would not render at all during development. `renderHTML` accepts a custom Html Container in its argument signature, but the calling function does not even attempt to find and pass the custom component in. But that is a secondary concern to the original post.
</details><br/>

If this is an acceptable proposal, I have no problem submitting a Pull Request, just let me know!

**Comment by MoOx** [comment]
Thanks for this detailed explanation @dlindahl !

- I want the HTML container to be used for dev server & static rendering indeed.
- Sure we can improve the situation for this kind of problem. I guess we should start indeed by not render BEFORE calling Html component. I think the thing we should do is to pass a function instead of props (body, script...) that will render, then returns this things.

Looking forward for your PR :)

**Comment by MoOx** [comment]
Hey guys, I changed stuffs required for Html components in https://github.com/phenomic/phenomic/pull/1138/commits/d4157fa4afadbd500a2e052dfa227262ee445d8c

You can take a look at what it can be used with styled-components https://github.com/phenomic/phenomic/pull/1138/commits/6b2f38a66be1d5a5e263032b51800edaa1725020

I will add more examples, but should be pretty similar. Tell me how it looks!
